src/Model.integration.test.ts

```diff
@@ -0,0 +1,244 @@
+
+    const sync = model.sync({
+      data: {
+        foo: 'foobar',
+        bar: {
+          baz: 1,
+        },
+      },
+      sequenceId: '1',
+    });
+    expect(model.state).toEqual('syncing');
+
+    await sync;
+    expect(model.state).toEqual('ready');
+
+    await model.pause();
+    expect(model.state).toEqual('paused');
+
+    await model.resume();
+    expect(model.state).toEqual('ready');
+
+    await model.dispose();
+    expect(model.state).toEqual('disposed');
+  });
+
+  it<TestContext>('successfully sets the data from the event in optimistic()', async ({
+    ably,
+    model,
+    channelName,
+    eventData,
+  }) => {
+    await model.sync(eventData.data);
+
+    let subscription = new Subject<void>();
+    const subscriptionCalls = getEventPromises(subscription, 3);
+    const subscriptionSpy = vi.fn(() => subscription.next());
+
+    model.subscribe(subscriptionSpy);
+
+    const [confirmation] = await model.optimistic(eventData);
+
+    expect(model.data.optimistic).toEqual(eventData.data);
+
+    const channel = ably.channels.get(channelName);
+    await channel.publish({
+      data: eventData.data,
+      name: 'update',
+      extras: {
+        headers: {
+          'x-ably-models-event-uuid': eventData.mutationId,
+        },
+      },
+    });
+
+    await subscriptionCalls[2];
+    await confirmation;
+
+    expect(subscriptionSpy).toHaveBeenCalledTimes(3);
+    expect(subscriptionSpy).toHaveBeenNthCalledWith(3, null, eventData.data);
+    expect(model.data.confirmed).toEqual(eventData.data);
+  });
+
+  it<TestContext>('rejects the data in optimistic() and rolls back the changes if back-end rejects the data', async ({
+    ably,
+    model,
+    channelName,
+    eventData,
+  }) => {
+    await model.sync(eventData.data);
+    let subscription = new Subject<void>();
+    const subscriptionCalls = getEventPromises(subscription, 3);
+    const subscriptionSpy = vi.fn(() => subscription.next());
+    model.subscribe(subscriptionSpy);
+
+    await model.optimistic(eventData);
+    expect(model.data.optimistic).toEqual(eventData.data);
+    const channel = ably.channels.get(channelName);
+    await channel.publish({
+      data: eventData.data,
+      name: 'update',
+      extras: {
+        headers: {
+          'x-ably-models-reject': true,
+        },
+      },
+    });
+
+    await subscriptionCalls[1];
+
+    expect(subscriptionSpy).toHaveBeenCalledTimes(2);
+    expect(subscriptionSpy).toHaveBeenNthCalledWith(2, null, eventData.data);
+    expect(model.data.confirmed).toEqual(undefined);
+  });
+
+  it<TestContext>('rejects the data and rolls back the changes if optimistic() timeouts', async ({
+    model,
+    eventData,
+  }) => {
+    await model.sync(eventData.data);
+    const [confirmation] = await model.optimistic(eventData, { timeout: 10 });
+    expect(model.data.optimistic).toEqual(eventData.data);
+
+    try {
+      await confirmation;
+    } catch (e) {
```

Comment:
This test doesn't assert that `confirmation` rejects, tests will pass if it resolves as the catch block is never run

Reply:
You can use `await expect(promise).rejects.toThrow()`

Reply:
This test should really be the same as the one above, i.e. it still asserts that the `subscribe` callback is invoked correctly